### DISCUSSION TITLE
请教下 bgfx.set_transform() 的使用问题

### DISCUSSION THREAD

**Discussion by constly** [ejoy/ant#167] [Q&A]
我最近在学习bgfx，初步想法是把部分bgfx的example用ant实现一遍（参考的 ant/test/native_bgfx）  
目前在做到第01-cubes时遇到点问题：没搞明白应该如何给 bgfx.set_transform()  传参数（没在引擎lua中找到使用参考）。

原生c++代码如下（文件[cubes.cpp](https://github.com/bkaradzic/bgfx/blob/master/examples/01-cubes/cubes.cpp)317行）
```c++
float mtx[16];
bx::mtxRotateXY(mtx, time + xx*0.21f, time + yy*0.37f);
mtx[12] = -15.0f + float(xx)*3.0f;
mtx[13] = -15.0f + float(yy)*3.0f;
mtx[14] = 0.0f;

// Set model matrix for rendering.
bgfx::setTransform(mtx);
```
想请教下，这段代码如何用ant lua实现
(我也注意到云大的这个库[lua-bgfx](https://github.com/cloudwu/lua-bgfx)，但我感觉差异有点大，不太好参考= =)

**Comment by actboy168** [comment]
bgfx.set_transform接受一个由math3d生成的id，也就是一个矩阵。

**Reply by constly** [reply]
噢，我明白了，是这样这样吧！  我这里代码有点问题，渲染不出来，我等会尝试下看看
```
local ref1 = math3d.ref(math3d.matrix(
        1.0, 0.0, 0.0, 0.0,
        0.0, 1.0, 0.0, 0.0,
        0.0, 0.0, 1.0, 0.0,
        0, 0, 0, 0.0))
bgfx.set_transform(ref1)
```

**Reply by actboy168** [reply]
对但不完全对。math3d有三种类型的id
1. 临时id可以存活两帧
2. 用math3d.mark可以把临时id变成永久id，在被unmark前不会释放
3. 用math3d.ref可以把临时id变成使用userdata管理，ud存活时不会释放

你这里用临时id就行，也就是可以直接把math3d.ref删了

**Comment by cloudwu** [comment]
lua-bgfx 这个库和 ant 用的完全一致。但是，`bgfx.set_transform()` 这样的 api 当需要传矩阵时，都是接收的 lightuserdata 。

ant 使用了 math3d ，矩阵是一个 id 。所以这里需要做一个桥接工作。在 lua-bgfx 那个独立库中，桥接是在 https://github.com/cloudwu/lua-bgfx/blob/master/mathadapter.lua 里完成的。

```lua
bgfx.set_transform = adapter.matrix(bgfx.set_transform, 1, 0)
```

这一行修改了这个函数，把第一个参数从 id 转换为了矩阵指针。

在 Ant 中，同样有代码进行转换：

https://github.com/ejoy/ant/blob/master/pkg/ant.math/adapter.lua

如果你不想通过 adapter 转换，那么可以使用 `math3d.value_ptr` 把 id 对应的指针取出来，这是  `ant/test/native_bgfx` 使用的方法。可以参考：https://github.com/ejoy/ant/blob/master/test/native_bgfx/pkg/ant.test.native_bgfx/init_system.lua#L262

**Reply by constly** [reply]
原来如此，我感觉lua-bgfx更适合参考学习，我后面直接参考这个得了~
